packages/techdocs-common/src/stages/publish/openStackSwift.ts

```diff
@@ -0,0 +1,260 @@
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+  }
+
+  async fetchTechDocsMetadata(
+    entityName: EntityName,
+  ): Promise<TechDocsMetadata> {
+    try {
+      return await new Promise<TechDocsMetadata>(async (resolve, reject) => {
+        const entityRootDir = `${entityName.namespace}/${entityName.kind}/${entityName.name}`;
+
+        const stream = this.storageClient.download({
+          container: this.containerName,
+          remote: `${entityRootDir}/techdocs_metadata.json`,
+        });
+
+        try {
+          const techdocsMetadataJson = await streamToBuffer(stream);
+          if (!techdocsMetadataJson) {
+            throw new Error(
+              `Unable to parse the techdocs metadata file ${entityRootDir}/techdocs_metadata.json.`,
+            );
+          }
+
+          const techdocsMetadata = JSON5.parse(
+            techdocsMetadataJson.toString('utf-8'),
+          );
+
+          resolve(techdocsMetadata);
+        } catch (err) {
+          this.logger.error(err.message);
+          reject(new Error(err.message));
+        }
+      });
+    } catch (e) {
+      throw new Error(`TechDocs metadata fetch failed, ${e.message}`);
+    }
+  }
+
+  /**
+   * Express route middleware to serve static files on a route in techdocs-backend.
+   */
+  docsRouter(): express.Handler {
+    return async (req, res) => {
+      // Trim the leading forward slash
+      // filePath example - /default/Component/documented-component/index.html
+
+      const filePath = req.path.replace(/^\//, '');
+
+      // Files with different extensions (CSS, HTML) need to be served with different headers
+      const fileExtension = path.extname(filePath);
+      const responseHeaders = getHeadersForFileExtension(fileExtension);
+
+      const stream = this.storageClient.download({
+        container: this.containerName,
+        remote: filePath,
+      });
+
+      try {
+        // Inject response headers
+        for (const [headerKey, headerValue] of Object.entries(
+          responseHeaders,
+        )) {
+          res.setHeader(headerKey, headerValue);
+        }
+
+        res.send(await streamToBuffer(stream));
+      } catch (err) {
+        this.logger.warn(err.message);
+        res.status(404).send(err.message);
+      }
+    };
+  }
+
+  /**
+   * A helper function which checks if index.html of an Entity's docs site is available. This
+   * can be used to verify if there are any pre-generated docs available to serve.
+   */
+  async hasDocsBeenGenerated(entity: Entity): Promise<boolean> {
+    try {
+      const entityRootDir = `${entity.metadata.namespace}/${entity.kind}/${entity.metadata.name}`;
+
+      return new Promise(res => {
+        this.storageClient.getFile(
+          this.containerName,
+          `${entityRootDir}/index.html`,
+          (err: any, file: any) => {
+            if (!err && file) {
+              res(true);
+            } else res(false);
```

Comment:
this one should probably be more robust with actual errors, right? so that they propagate as a rejection to the caller of hasDocsBeenGenerated

Reply:
@freben Hello I apologize for the late reply. I added warn log for this line. i don't understand what is your mean can you explain?

Reply:
I meant that you probably should not log at all. Instead do `if (err) { rej(err) }` so that the caller gets informed that there's an error because that's different from it not being present. Right?

Reply:
First of all, I check the `PublisherBase` class `hasDocsBeenGenerated` method should return a boolean instead of an error object. Here implementation; 
https://github.com/backstage/backstage/blob/4117d00f2177970a06f2ff4126cfb0a65cfdfcfd/packages/techdocs-common/src/stages/publish/types.ts#L75

When the requested file does not exist in the OpenStack Swift container `pkgcloud` returns a 404 error object. I think it's not an actual error because 404 meaning not found for this reason we call `res(false)`

Check google and s3 integrations they do the same thing, always return a boolean.
S3: https://github.com/backstage/backstage/blob/ae30b916df48cb554f171c04791bdc85be1a201a/packages/techdocs-common/src/stages/publish/awsS3.ts#L269

Google:
https://github.com/backstage/backstage/blob/ae30b916df48cb554f171c04791bdc85be1a201a/packages/techdocs-common/src/stages/publish/googleStorage.ts#L209